File: app/udptransport/src/main/java/org/exthmui/share/udptransport/UDPSender.java
package org.exthmui.share.udptransport;

import android.content.Context;
import android.util.Log;
import android.util.Pair;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.StringUtils;
import org.exthmui.share.shared.base.Entity;
import org.exthmui.share.shared.base.FileInfo;
import org.exthmui.share.shared.base.receive.SenderInfo;
import org.exthmui.share.shared.base.results.SuccessTransmissionResult;
import org.exthmui.share.shared.base.results.TransmissionResult;
import org.exthmui.share.shared.base.send.ReceiverInfo;
import org.exthmui.share.shared.exceptions.trans.ReceiverCancelledException;
import org.exthmui.share.shared.exceptions.trans.SenderCancelledException;
import org.exthmui.share.shared.exceptions.trans.TimedOutException;
import org.exthmui.share.shared.exceptions.trans.TransmissionException;
import org.exthmui.share.udptransport.packets.AbstractCommandPacket;
import org.exthmui.share.udptransport.packets.FilePacket;
import org.exthmui.share.udptransport.packets.ResendRequestPacket;
import org.exthmui.utils.StackTraceUtils;

import java.io.BufferedInputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.Type;
import java.net.DatagramPacket;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketAddress;
import java.net.SocketException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.Future;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/**
 * UDPSender
 * <p>
 * initialize -> sendAsync
 */
public class UDPSender {
    public static final String TAG = "UDPSender";
    private final Object lock = new Object();

    @NonNull
    private final Context context;
    @NonNull
    private final SendingListener listener;

    @Nullable
    private TCPUtil tcpUtil;
    @Nullable
    private UDPUtil udpUtil;
    @Nullable
    private UDPUtil.Handler handler;

    long totalBytesToSend = 0;
    long bytesSent = 0;
    ReceiverInfo receiverInfo = null;
    FileInfo[] fileInfos = null;
    private int remoteUdpPort;
    private byte connId;

    private boolean canceled;
    private boolean remoteCanceled;

    private final ThreadPoolExecutor coreThreadPool = new ThreadPoolExecutor(1,
            1, 0L, TimeUnit.SECONDS, new SynchronousQueue<>(),
            r -> new Thread(r, r.toString()));
    private final ThreadPoolExecutor threadPool = new ThreadPoolExecutor(0,
            2, 1L, TimeUnit.SECONDS, new SynchronousQueue<>(),
            r -> new Thread(r, r.toString()));

    public UDPSender(@NonNull Context context, @NonNull SendingListener listener) {
        this.context = context;
        this.listener = listener;
    }

    public void cancel() {
        synchronized (lock) {
            canceled = true;
            lock.notifyAll();
        }
    }

    public void initialize() throws SocketException {
        tcpUtil = new TCPUtil();
        udpUtil = new UDPUtil();
        udpUtil.setTAG(TAG);
        udpUtil.setStateChecker(this::checkCanceled);
    }

    public Future<Integer> sendAsync(@NonNull Entity[] entities, @NonNull FileInfo[] fileInfos,
                                     @NonNull SenderInfo sender, @NonNull SocketAddress tcpAddress) {
        if (entities.length != fileInfos.length) throw new IllegalArgumentException();
        return coreThreadPool.submit(() -> {
            try {
                send(entities, fileInfos, sender, tcpAddress);
            } catch (Throwable tr) {
                Log.e(TAG, String.format("Error occured while sending: %s(message: %s)", tr, tr.getMessage()));
                Log.e(TAG, StackTraceUtils.getStackTraceString(tr.getStackTrace()));
                listener.onComplete(
                        org.exthmui.share.shared.misc.Constants.TransmissionStatus.ERROR.getNumVal(),
                        null);
            } finally {
                releaseResources();
            }
            return null;//TODO
        });
    }

    private void send(@NonNull Entity[] entities, @NonNull FileInfo[] fileInfos,
                      @NonNull SenderInfo sender, @NonNull SocketAddress tcpAddress) throws IOException {
        if (entities.length != fileInfos.length) throw new IllegalArgumentException();
        assert tcpUtil != null;
        tcpUtil.connect(tcpAddress);// (1)
        tcpUtil.writeJson(sender);// (2)
        tcpUtil.writeJson(fileInfos);// (3)
        listener.onProgressUpdate(
                org.exthmui.share.shared.misc.Constants.TransmissionStatus.WAITING_FOR_ACCEPTATION.getNumVal(),
                0, 0, null, 0, 0);
        String[] accepted = tcpUtil.readJson(String[].class);// (4)
        if (accepted == null || accepted.length == 0) {
            Log.d(TAG, "No file were accepted");
            listener.onComplete(
                    org.exthmui.share.shared.misc.Constants.TransmissionStatus.REJECTED.getNumVal(),
                    null);
            return;
        } else listener.onAccepted(accepted);
        List<Entity> entitiesToSend = new ArrayList<>();
        List<FileInfo> fileInfosToSend = new ArrayList<>();
        List<String> acceptedIdsAsList = Arrays.asList(accepted);
        for (FileInfo fileInfo : fileInfos) {
            if (acceptedIdsAsList.contains(fileInfo.getId())) {
                totalBytesToSend += fileInfo.getFileSize();
                fileInfosToSend.add(fileInfo);
                entitiesToSend.add(entities[Arrays.asList(fileInfos).indexOf(fileInfo)]);
            }
        }

        tcpUtil.setCommandListener(cmd -> {
            if (cmd.equals(Constants.COMMAND_CANCEL)) {
                remoteCanceled = true;
                synchronized (lock) {
                    lock.notifyAll();
                }
            } else if (StringUtils.startsWith(cmd, Constants.COMMAND_UDP_SOCKET_READY)) { // e.g. UDP_READY5000:-128
                    remoteUdpPort = Integer.parseInt(cmd.replace(Constants.COMMAND_UDP_SOCKET_READY, "").split(":")[0]);
                    connId = Byte.parseByte(cmd.replace(Constants.COMMAND_UDP_SOCKET_READY, "").split(":")[1]);
            }
        });// (4-6)
        while (remoteUdpPort == 0) {// (6)
            if (checkCanceled()) return;
            synchronized (lock) {
                try {
                    lock.wait();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
        assert udpUtil != null;
        udpUtil.addHandler(context, connId);
        handler = udpUtil.getHandler(connId);
        udpUtil.connect(new InetSocketAddress(tcpUtil.getInetAddress(), remoteUdpPort));
        udpUtil.startListening();
        tcpUtil.writeCommand(String.format(Locale.ROOT, "%s%d:%d", Constants.COMMAND_UDP_SOCKET_READY, udpUtil.getLocalPort(), connId));// (6)
        listener.onProgressUpdate(
                org.exthmui.share.shared.misc.Constants.TransmissionStatus.CONNECTION_ESTABLISHED.getNumVal(),
                totalBytesToSend, 0, null, 0, 0);
        for (Entity entity : entitiesToSend) {
            Log.d(TAG, "Start sending file: " + entity.getFileName());
            sendFile(entity, fileInfosToSend.get(entitiesToSend.indexOf(entity)));

            if (checkCanceled()) return;
        }
        @SuppressWarnings("unchecked") Map<String, Pair<Integer, String>> resultMap =
                tcpUtil.readJson(Map.class, new TypeToken<Map<String, Pair<Integer, String>>>() {
                }.getType());// (16)
        for (String id : resultMap.keySet()) {
            if ((Objects.requireNonNull(resultMap.get(id)).first &
                    org.exthmui.share.shared.misc.Constants.TransmissionStatus.ERROR.getNumVal()) !=
                    Objects.requireNonNull(resultMap.get(id)).first) {
                listener.onComplete(org.exthmui.share.shared.misc.Constants.TransmissionStatus.REMOTE_ERROR.getNumVal(),
                        resultMap);
                return;
            }
        }
        listener.onComplete(
                org.exthmui.share.shared.misc.Constants.TransmissionStatus.COMPLETED.getNumVal(),
                resultMap);
    }

    public TransmissionResult sendFile(Entity entity, FileInfo fileInfo) throws IOException {
        try {
            listener.onProgressUpdate(
                    org.exthmui.share.shared.misc.Constants.TransmissionStatus.CONNECTION_ESTABLISHED.getNumVal(),
                    totalBytesToSend, 0, null, 0, 0);
            InputStream stream = entity.getInputStream(context);
            BufferedInputStream inputStream = stream instanceof BufferedInputStream ?
                    (BufferedInputStream) stream : new BufferedInputStream(stream);
            assert udpUtil != null;
            assert handler != null;
            FilePacket sendPacket = new FilePacket();
            final byte startGroup = Constants.START_GROUP_ID;
            final byte endGroup = Constants.END_GROUP_ID;
            final short startPacket = Constants.START_PACKET_ID;
            final short endPacket = Constants.END_PACKET_ID;
            byte curGroup = startGroup;
            short curPacket = startPacket;
            byte[][] bufTemp = new byte[Short.MAX_VALUE - Short.MIN_VALUE + 1][];
            byte[] buf = new byte[Constants.DATA_LEN_MAX_HI];
            int len = 0;
            handler.sendIdentifier(Constants.START_IDENTIFIER, Constants.START_ACK_IDENTIFIER, ArrayUtils.addFirst(ByteUtils.shortToBytes(startPacket), curGroup));// (7), (8)
            while ((len = (inputStream.read(buf))) > 0) {
                bufTemp[curPacket - Short.MIN_VALUE] = buf.clone();
                handler.sendPacket(sendPacket.setPacketId(curPacket).setGroupId(curGroup).setData(buf, len));// (9)

                if (checkCanceled())
                    return remoteCanceled ? new ReceiverCancelledException(context) :
                            new SenderCancelledException(context);

                if (curPacket == endPacket) {
                    boolean packetAllSent = false;
                    while (!packetAllSent) {
                        handler.sendIdentifier(Constants.END_IDENTIFIER, Constants.END_ACK_IDENTIFIER, ArrayUtils.addFirst(ByteUtils.shortToBytes(curPacket), curGroup));// (10), (11)

                        ResendRequestPacket resendReq = handler.receivePacket(ResendRequestPacket::new, Constants.ACK_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS, Constants.MAX_ACK_TRYOUTS);// (12)
                        short[] packetIds = resendReq.getPacketIds();
                        if (packetIds.length > 0) {
                            FilePacket[] packetsToResend = new FilePacket[packetIds.length];
                            for (int i = 0; i < packetIds.length; i++) {
                                packetsToResend[i] = new FilePacket().setGroupId(curGroup).setPacketId(packetIds[i]).setData(bufTemp[packetIds[i]]);
                            }
                            resendPackets(packetsToResend, curGroup, curPacket);// (13)
                        } else packetAllSent = true;
                    }

                    if (curGroup + 1 > endGroup) {
                        handler.sendIdentifier(Constants.GROUP_ID_RESET_IDENTIFIER, Constants.GROUP_ID_RESET_ACK_IDENTIFIER, new byte[]{curGroup, startGroup});
                        curGroup = startGroup;
                        handler.sendIdentifier(Constants.START_IDENTIFIER, Constants.START_ACK_IDENTIFIER, ArrayUtils.addFirst(ByteUtils.shortToBytes(startPacket), curGroup));// (7), (8)
                    } else {
                        curGroup++;
                        handler.sendIdentifier(Constants.START_IDENTIFIER, Constants.START_ACK_IDENTIFIER, ArrayUtils.addFirst(ByteUtils.shortToBytes(startPacket), curGroup));// (7), (8)
                    }
                    curPacket = startPacket;
                } else {
                    curPacket++;
                }
            }

            curPacket--;
            boolean packetAllSent = false;
            while (!packetAllSent) {
                handler.sendIdentifier(Constants.END_IDENTIFIER, Constants.END_ACK_IDENTIFIER, ArrayUtils.addFirst(ByteUtils.shortToBytes(curPacket), curGroup));// (10), (11)

                ResendRequestPacket resendReq = handler.receivePacket(ResendRequestPacket::new, Constants.ACK_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS, Constants.MAX_ACK_TRYOUTS);// (12)

                short[] packetIds = resendReq.getPacketIds();
                if (packetIds.length > 0) {
                    FilePacket[] packetsToResend = new FilePacket[packetIds.length];
                    for (int i = 0; i < packetIds.length; i++) {
                        if (bufTemp[packetIds[i] - Short.MIN_VALUE] == null) break;
                        packetsToResend[i] = new FilePacket().setGroupId(curGroup).setPacketId(packetIds[i]).setData(bufTemp[packetIds[i] - Short.MIN_VALUE]);
                        if (bufTemp[packetIds[i] - Short.MIN_VALUE].length != Constants.DATA_LEN_MAX_HI)
                            break;
                    }
                    resendPackets(packetsToResend, curGroup, curPacket);// (13)
                } else packetAllSent = true;
            }

            handler.sendIdentifier(Constants.FILE_END_IDENTIFIER, Constants.FILE_END_ACK_IDENTIFIER);// (7), (8)
            return new SuccessTransmissionResult(context);
        } catch (TransmissionException e) {
            return e;
        }
    }

    public void resendPackets(FilePacket[] packets, byte curGroup, short curGroupEndPacket) throws IOException, TimedOutException {
        assert handler != null;
        Log.d(TAG, String.format("Resend start: %s", Arrays.toString(Arrays.stream(packets).map(FilePacket::getPacketId).toArray())));
        handler.sendIdentifier(Constants.START_IDENTIFIER, Constants.START_ACK_IDENTIFIER, new byte[]{curGroup});
        for (FilePacket packet : packets)
            handler.sendPacket(packet);

        handler.sendIdentifier(Constants.END_IDENTIFIER, Constants.END_ACK_IDENTIFIER, ArrayUtils.addFirst(ByteUtils.shortToBytes(curGroupEndPacket), curGroup));// (10), (11)

        ResendRequestPacket resendReq = handler.receivePacket(ResendRequestPacket::new, Constants.ACK_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS, Constants.MAX_ACK_TRYOUTS);// (12)

        short[] packetIds = resendReq.getPacketIds();
        Log.d(TAG, String.format("Resend requested: %s", Arrays.toString(packetIds)));
        if (packetIds.length > 0) {
            ArrayList<FilePacket> packetsToResend = new ArrayList<>();
            for (short packetId : packetIds) {
                for (FilePacket packet : packets) {
                    if (packet.getPacketId() == packetId) {
                        packetsToResend.add(packet);
                        break;
                    }
                }

            }
            resendPackets(packetsToResend.toArray(new FilePacket[0]), curGroup, curGroupEndPacket);// (13)
        }
    }

    public void releaseResources() {
        if (tcpUtil != null) {
            tcpUtil.releaseResources();
            tcpUtil = null;
        }
        if (udpUtil != null) {
            udpUtil.releaseResources();
            udpUtil = null;
        }
        coreThreadPool.shutdown();
        threadPool.shutdown();
    }

    public boolean checkCanceled() {
        if (canceled) {
            Log.d(TAG, "Sending canceled by sender");
            listener.onComplete(org.exthmui.share.shared.misc.Constants.TransmissionStatus.SENDER_CANCELLED.getNumVal(),
                    null);
            return true;
        } else if (remoteCanceled) {
            Log.d(TAG, "Sending canceled by receiver");
            listener.onComplete(org.exthmui.share.shared.misc.Constants.TransmissionStatus.RECEIVER_CANCELLED.getNumVal(),
                    null);
            return true;
        }
        return false;
    }

    public interface SendingListener {
        void onAccepted(String[] fileIdsAccepted);

        void onProgressUpdate(int status, long totalBytesToSend, long bytesSent, String curFileId,
                              long curFileBytesToSend, long curFileBytesSent);

        void onComplete(int status, Map<String, Pair<Integer, String>> resultMap);
    }

    public interface PacketFactory<T extends AbstractCommandPacket<T>> {
        T produce(DatagramPacket datagramPacket);
    }
}

File: app/udptransport/src/main/java/org/exthmui/share/udptransport/UDPSender.kt
package org.exthmui.share.udptransport

import android.content.Context
import android.util.Log
import androidx.annotation.IntRange
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import org.apache.commons.lang3.ArrayUtils
import org.exthmui.share.shared.base.Entity
import org.exthmui.share.shared.base.FileInfo
import org.exthmui.share.shared.base.receive.SenderInfo
import org.exthmui.share.shared.base.results.BasicTransmissionResult
import org.exthmui.share.shared.base.results.SuccessTransmissionResult
import org.exthmui.share.shared.base.results.TransmissionResult
import org.exthmui.share.shared.base.send.ReceiverInfo
import org.exthmui.share.shared.exceptions.trans.CancelledException
import org.exthmui.share.shared.exceptions.trans.RejectedException
import org.exthmui.share.shared.exceptions.trans.RemoteErrorException
import org.exthmui.share.shared.exceptions.trans.SenderCancelledException
import org.exthmui.share.shared.exceptions.trans.TransmissionException
import org.exthmui.share.udptransport.packets.FilePacket
import org.exthmui.share.udptransport.packets.ResendRequestPacket
import org.exthmui.utils.StackTraceUtils
import java.io.BufferedInputStream
import java.io.IOException
import java.net.DatagramPacket
import java.net.InetSocketAddress
import java.util.Locale
import java.util.concurrent.TimeUnit
import kotlin.properties.Delegates

class UDPSender(context: Context, private val listener: SendingListener?) :
    ResultMapCompletable(context) {
    private val tcpUtil: TCPUtil by lazy {
        TCPUtil().apply {
            setTAG(TAG)
            registerCommandConsumer(Regex(Constants.COMMAND_CANCEL)) {
                isReceiverCanceled = true
                cancel()
                true
            }
        }
    }
    private val udpUtil: UDPUtil by lazy {
        UDPUtil().apply {
            setTAG(TAG)
            setStateChecker { runBlocking { isCanceled() } }
        }
    }
    private lateinit var handler: UDPUtil.Handler
    var totalBytesToSend: Long = 0
    var bytesSent: Long = 0
    var receiverInfo: ReceiverInfo? = null
    var fileInfos: Array<FileInfo>? = null
    override var isSenderCanceled: Boolean = false
    override var isReceiverCanceled: Boolean = false
    private var remoteUdpPort = CompletableDeferred<Int>()
    private var connId = CompletableDeferred<Byte>()

    private val udpReadyCommandConsumer: (String) -> Boolean = {
        // e.g. UDP_READY5000:-128
        remoteUdpPort.complete(
            it.replace(Constants.COMMAND_UDP_SOCKET_READY, "").split(":".toRegex())
                .dropLastWhile { it.isEmpty() }.toTypedArray()[0].toInt()
        )
        connId.complete(
            it.replace(Constants.COMMAND_UDP_SOCKET_READY, "").split(":".toRegex())
                .dropLastWhile { it.isEmpty() }.toTypedArray()[1].toByte()
        )
        true
    }

    private lateinit var job: Deferred<Pair<TransmissionResult, Map<String, TransmissionResult>>>

    fun cancel() {
        isSenderCanceled = true
        job.cancel()
    }

    private fun updateProgress(
        status: Int,
        totalBytesToSend: Long,
        bytesSent: Long,
        curFileId: String?,
        curFileBytesToSend: Long,
        curFileBytesSent: Long
    ) {
        listener?.onProgressUpdate(
            status, totalBytesToSend, bytesSent, curFileId, curFileBytesToSend, curFileBytesSent
        )
    }

    private fun accept(accepted: Array<String>) {
        listener?.onAccepted(accepted)
    }

    fun sendAsync(
        entities: Array<Entity>,
        fileInfos: Array<FileInfo>,
        sender: SenderInfo,
        tcpAddress: InetSocketAddress
    ): Deferred<Pair<TransmissionResult, Map<String, TransmissionResult>>> {
        require(entities.size == fileInfos.size)
        job = ProcessLifecycleOwner.get().lifecycleScope.async(Dispatchers.Default) {
            val result = try {
                send(entities, fileInfos, sender, tcpAddress)
            } catch (tr: Throwable) {
                Log.e(
                    String.format(UDPSender.TAG, connId.await()),
                    "Error occurred while sending: $tr(message: ${tr.message})\n" + StackTraceUtils.getStackTraceString(
                        tr.stackTrace
                    )
                )
                setRestAs(SenderCancelledException(context))
                complete()
            } finally {
                releaseResources()
            }

            listener?.onComplete(result.first, result.second)
            result
        }
        return job
    }

    @Throws(IOException::class)
    private suspend fun send(
        entities: Array<Entity>,
        fileInfos: Array<FileInfo>,
        sender: SenderInfo,
        tcpAddress: InetSocketAddress
    ): Pair<TransmissionResult, Map<String, TransmissionResult>> = coroutineScope {
        require(entities.size == fileInfos.size)
        tcpUtil.connect(tcpAddress) // (1)
        tcpUtil.writeJson(sender) // (2)
        tcpUtil.writeJson(fileInfos) // (3)
        updateProgress(
            org.exthmui.share.shared.misc.Constants.TransmissionStatus.WAITING_FOR_ACCEPTATION.numVal,
            0,
            0,
            null,
            0,
            0
        )
        val accepted: Array<String> = tcpUtil.readJson(Array<String>::class.java) // (4)
        Log.d(TAG, "Remote accepted: ${ArrayUtils.toString(accepted)}")
        for (key in accepted) mutableResultMap[key] = BasicTransmissionResult.UNKNOWN_RESULT
        if (accepted.isEmpty()) {
            Log.d(TAG, "No file were accepted")
            return@coroutineScope completeRestWith(RejectedException(context))
        } else accept(accepted)
        val entitiesToSend: MutableList<Entity> = ArrayList()
        val fileInfosToSend: MutableList<FileInfo> = ArrayList()
        val acceptedIdsAsList = listOf(*accepted)
        for (fileInfo in fileInfos) {
            if (acceptedIdsAsList.contains(fileInfo.id)) {
                totalBytesToSend += fileInfo.fileSize
                fileInfosToSend.add(fileInfo)
                entitiesToSend.add(entities[listOf(*fileInfos).indexOf(fileInfo)])
            }
        }

        tcpUtil.readCommand(Regex(Constants.COMMAND_UDP_SOCKET_READY + ".+"), true).await().let {
            // e.g. UDP_READY5000:-128
            val params = it.replace(Constants.COMMAND_UDP_SOCKET_READY, "").split(":".toRegex())
                .dropLastWhile { it.isEmpty() }.toTypedArray()
            remoteUdpPort.complete(params[0].toInt())
            connId.complete(params[1].toByte())
            true
        }// (4-6)

        // TODO:CancellationException Handle
        remoteUdpPort.await()// (6)
        checkCanceledAndComplete()?.let { return@coroutineScope it }

        handler = udpUtil.addHandler(context, connId.await())
        udpUtil.connect(InetSocketAddress(tcpUtil.inetAddress, remoteUdpPort.await()))
        udpUtil.startListening()
        tcpUtil.writeCommand(
            String.format(
                Locale.ROOT,
                "%s%d:%d",
                Constants.COMMAND_UDP_SOCKET_READY,
                udpUtil.localPort,
                connId.await()
            )
        ) // (6)
        updateProgress(
            org.exthmui.share.shared.misc.Constants.TransmissionStatus.CONNECTION_ESTABLISHED.numVal,
            totalBytesToSend,
            0,
            null,
            0,
            0
        )
        for (entity in entitiesToSend) {
            Log.d(TAG, "Start sending file: " + entity.fileName)
            sendFile(entity, fileInfosToSend[entitiesToSend.indexOf(entity)])
            checkCanceledAndComplete()?.let { return@coroutineScope it }
        }
        val resultMap: Map<String, TransmissionResult> =
            TransmissionResult.stringToMap(context, tcpUtil.readBare()) // (16)

        resultMap.forEach {
            if (mutableResultMap.containsKey(it.key)) if (it.value is TransmissionException && it.value !is CancelledException && it.value !is RemoteErrorException) mutableResultMap[it.key] =
                RemoteErrorException(context, it.value as TransmissionException)
        }

        for (id in resultMap.keys) {
            if (org.exthmui.share.shared.misc.Constants.TransmissionStatus.ERROR.numVal != resultMap[id]?.statusCode) {
                complete(
                    RemoteErrorException(context)
                )
            }
        }

        complete()
    }

    enum class GroupStatus {
        Waiting,// Before START
        Transferring, Ending,// Packets all sent out
        DoResendCheck// After END, before resending check
    }

    enum class FileStatus {
        Waiting, Transferring, GroupIdExceeded, Ended
    }

    @Throws(IOException::class)
    suspend fun sendFile(entity: Entity, fileInfo: FileInfo?): TransmissionResult = coroutineScope {
        try {
            updateProgress(
                org.exthmui.share.shared.misc.Constants.TransmissionStatus.CONNECTION_ESTABLISHED.numVal,
                totalBytesToSend,
                0,
                null,
                0,
                0
            )
            val stream = entity.getInputStream(context)
            val inputStream =
                if (stream is BufferedInputStream) stream else BufferedInputStream(stream)
            val sendPacket = FilePacket()

            val startGroup = Constants.START_GROUP_ID
            val endGroup = Constants.END_GROUP_ID
            val startPacket = Constants.START_PACKET_ID
            val endPacket = Constants.END_PACKET_ID

            var isResending = false
            var groupStatus = GroupStatus.Waiting
            var fileStatus = FileStatus.Waiting

            @IntRange(
                from = Byte.MIN_VALUE.toLong(),
                to = Byte.MAX_VALUE.toLong()
            ) var curGroup by Delegates.vetoable(startGroup.toInt()) { _, _, newValue ->
                if (newValue > endGroup) {
                    fileStatus = FileStatus.GroupIdExceeded
                    true
                } else true
            }

            @IntRange(
                from = Byte.MIN_VALUE.toLong(),
                to = Byte.MAX_VALUE.toLong()
            ) var curPacket by Delegates.vetoable(startPacket.toInt()) { _, _, newValue ->
                if (newValue > endPacket) {
                    groupStatus = GroupStatus.Ending
                    false
                } else true
            }

            var packetIdIterator: Iterator<Number> by Delegates.observable((startPacket..endPacket).iterator()) { _, _, _ ->
                groupStatus = GroupStatus.Waiting
            }

            fun resetIdIterator() {
                packetIdIterator = (startPacket..endPacket).iterator()
            }

            val bufTemp = arrayOfNulls<ByteArray>(endPacket - startPacket + 1)
            val bufLen = arrayOfNulls<Int>(bufTemp.size)
            val buf = ByteArray(Constants.DATA_LEN_MAX_HI)
            var len: Int
            while (true) {
                checkCanceledSingleResult()?.let { return@coroutineScope it }
                when (fileStatus) {
                    FileStatus.Waiting -> {
                        fileStatus = FileStatus.Transferring
                    }

                    FileStatus.GroupIdExceeded -> {
                        handler.sendIdentifier(
                            Constants.Identifier.GROUP_ID_RESET,
                            byteArrayOf(curGroup.toByte(), startGroup)
                        )
                        curGroup = startGroup.toInt()
                    }

                    FileStatus.Transferring -> do {
                        when (groupStatus) {
                            GroupStatus.Waiting -> {
                                handler.sendIdentifier(
                                    Constants.Identifier.START, ArrayUtils.addFirst(
                                        ByteUtils.shortToBytes(startPacket), curGroup.toByte()
                                    )
                                ) // (7), (8)
                                groupStatus = GroupStatus.Transferring
                            }

                            GroupStatus.Transferring -> {
                                if (packetIdIterator.hasNext())
                                    curPacket = packetIdIterator.next().toInt()
                                else {
                                    groupStatus = GroupStatus.Ending
                                    continue
                                }
                                if (!isResending) {
                                    if (inputStream.read(buf).also { len = it } < 0) {
                                        groupStatus = GroupStatus.Ending
                                        fileStatus = FileStatus.Ended
                                        continue
                                    }
                                    bufTemp[curPacket - startPacket] =
                                        buf //TODO:.clone() needed?
                                    bufLen[curPacket - startPacket] =
                                        len
                                }
                                handler.sendPacket(
                                    sendPacket.setPacketId(curPacket.toShort())
                                        .setGroupId(curGroup.toByte())
                                        .setData(
                                            bufTemp[curPacket - startPacket],
                                            bufLen[curPacket - startPacket]!!
                                        )
                                ) // (9)
                            }

                            GroupStatus.Ending -> {
                                handler.sendIdentifier(
                                    Constants.Identifier.END, ArrayUtils.add(
                                        ArrayUtils.addFirst(
                                            ByteUtils.shortToBytes(curPacket.toShort()),
                                            curGroup.toByte()
                                        ),
                                        (if (isResending) 0x1 else 0x0).toByte()
                                    )
                                ) // (10), (11)
                                groupStatus = GroupStatus.DoResendCheck
                            }

                            GroupStatus.DoResendCheck -> {
                                val resendReq = handler.receivePacket<ResendRequestPacket>(
                                    { packet: DatagramPacket? ->
                                        ResendRequestPacket(packet)
                                    },
                                    Constants.ACK_TIMEOUT_MILLIS,
                                    TimeUnit.MILLISECONDS,
                                    Constants.MAX_ACK_TRYOUTS
                                ) // (12)
                                val packetIds = resendReq.packetIds
                                if (packetIds.isNotEmpty()) {
                                    Log.d(TAG, "Resend needed for group $curGroup: $packetIds")
                                    isResending = true
                                    packetIdIterator = packetIds.iterator()
                                } else {
                                    Log.d(TAG, "No resend needed for group $curGroup")
                                    isResending = false
                                    curGroup++
                                    resetIdIterator()
                                }
                            }
                        }
                    } while (groupStatus != GroupStatus.Waiting)

                    FileStatus.Ended -> {
                        handler.sendIdentifier(Constants.Identifier.FILE_END) // (7), (8)
                        break
                    }
                }
            }
            SuccessTransmissionResult(context)
        } catch (e: TransmissionException) {
            e
        }
    }

    fun releaseResources() {
        tcpUtil.releaseResources()
        udpUtil.releaseResources()
    }

    interface SendingListener {
        fun onAccepted(fileIdsAccepted: Array<String>?)
        fun onProgressUpdate(
            status: Int,
            totalBytesToSend: Long,
            bytesSent: Long,
            curFileId: String?,
            curFileBytesToSend: Long,
            curFileBytesSent: Long
        )

        fun onComplete(status: TransmissionResult, resultMap: Map<String, TransmissionResult>?)
    }

    companion object {
        const val TAG = "UDPSender"
    }
}
